api/auth_middleware_test.go

```diff
@@ -1,63 +1,101 @@
 package api_test
 
 import (
+	b64 "encoding/base64"
 	"fmt"
 	"net/http"
 	"testing"
 
+	"github.com/topfreegames/Will.IAM/models"
 	helpers "github.com/topfreegames/Will.IAM/testing"
 )
 
-func TestAuthMiddlewareKeyPairShouldAuthenticateUser(t *testing.T) {
+func TestAuthMiddleware(t *testing.T) {
 	helpers.CleanupPG(t)
 
-	rootSA := helpers.CreateRootServiceAccountWithKeyPair(t)
-	app := helpers.GetApp(t)
-	req, err := http.NewRequest(http.MethodGet, "/service_accounts", nil)
-
-	if err != nil {
-		t.Errorf("Could not create HTTP request for /service_accounts")
-	}
-
-	req.Header.Set("Authorization", fmt.Sprintf(
-		"keypair %s:%s", rootSA.KeyID, rootSA.KeySecret,
-	))
-
-	response := helpers.DoRequest(t, req, app.GetRouter())
-
-	if response.Code != http.StatusOK {
-		t.Errorf("Expected status %d. Got %d", http.StatusOK, response.Code)
-	}
+	oauthSA := helpers.CreateRootServiceAccountWithOAuth(t, "oauthUser", "oauth.user@test.com")
+	keyPairSA := helpers.CreateRootServiceAccountWithKeyPair(t, "keyPairUser", "keypair.user@test.com")
+	tokensRepo := helpers.GetRepo(t).Tokens
+	tokens, _ := tokensRepo.FindByEmail(oauthSA.Email)
+	token := tokens[0]
+	emptyMap := make(map[string]string)
 
-	serviceAccountName := response.Header().Get("x-service-account-name")
-	if serviceAccountName != "root" {
-		t.Errorf("Expected service account name %s. Got %s", "root", serviceAccountName)
+	testCases := []struct {
+		testName                string
+		serviceAccount          *models.ServiceAccount
+		requestHeaders          map[string]string
+		expectedResponseHeaders map[string]string
+		expectedResponseCode    int
+	}{
+		{
+			testName:                "KeyPairAuthorization",
+			serviceAccount:          keyPairSA,
+			requestHeaders:          map[string]string{"authorization": fmt.Sprintf("keypair %s:%s", keyPairSA.KeyID, keyPairSA.KeySecret)},
+			expectedResponseHeaders: map[string]string{"x-service-account-name": "keyPairUser"},
+			expectedResponseCode:    http.StatusOK,
+		},
+		{
+			testName:                "OAuthAuthorization",
+			serviceAccount:          oauthSA,
+			requestHeaders:          map[string]string{"authorization": fmt.Sprintf("bearer %s", token.AccessToken)},
+			expectedResponseHeaders: map[string]string{"x-email": oauthSA.Email},
+			expectedResponseCode:    http.StatusOK,
+		},
+		{
+			testName:                "WrongOAuthTokenAuthorization",
```

Comment:
Suggestion: maybe `oauth_invalidContent` (same for others, eg, "oauth_ok", "keypair_ok", "keypair_invalidContent", etc).